feat: add jaeger uber-trace-id header propagation

tcolgate opened the pull request:
Jaeger uses a custom header for propagating span context. They are intending
to move to the w3c spec, but there is no clear time frame on that, lots of
existinng things using jaeger tracing will need this for now.

tcolgate commented:
There seems to be some level of dissatisfaction with the idea of having this in tree.

kjin commented:
I wasn't aware of the discussion in the Python repo, thanks for the context.

I currently can't spare the bandwidth to review the code yet. Sorry for the inconvenience -- I hope to get to it soon!

tcolgate commented:
@kjin no probs. Do you have any (very approximate), time frame for the next likely release? We were going to wait on that rather than publishing a version of this package internally.

kjin commented:
I'll be able to answer that question later today. If you don't mind, please ping again if I don't have an answer by tomorrow.

kjin commented:
@tcolgate I can release the module on its own once I merge it, actually. That being said, I'll need to get confirmation from the OC team given the resistance in the Python thread.

tcolgate commented:
@kjin we're also waiting on the fixes for the jaeger exporter, and http tracing. I think the current release also has the issue where get isn't wrapped in https? (though that is already fixed on master). We more or less need a release of -core -node -exporter-jaeger and this. I don't mean that to sound pushy, that's just my "go live" list, the time frame is all down to you of course.

re: the python thing, the spec says alternate propagations can be supported. Seems odd to say that and then not actually accept any. Of course, if the decision goes against this, we can always publish separately.

kjin commented:
@tcolgate Ok, thanks for the heads up.

kjin commented on packages/opencensus-propagation-jaeger/test/test-jaeger-format.ts:
```diff
@@ -0,0 +1,91 @@
+const jaegerFormat = new JaegerFormat();
+
+describe('JaegerPropagation', () => {
+  /** Should get the singleton trancing instance. */
```
This comment seems like it doesn't apply, right?

kjin commented on packages/opencensus-propagation-jaeger/src/jaeger-format.ts:
```diff
@@ -0,0 +1,97 @@
+import * as crypto from 'crypto';
+import * as uuid from 'uuid';
+
+const UBER_TRACE_ID = 'uber-trace-id';
```
Is it more accurate to call this the "uber Trace ID" or the "Jaeger Trace ID"?

tcolgate replied:
We can't really change the header value, as the client libraries all assume uber-trace-id. Dropping the prefix seems reasonable.

kjin replied:
Thanks. I was referring to the variable name only.

kjin commented on packages/opencensus-propagation-jaeger/test/test-jaeger-format.ts:
```diff
@@ -0,0 +1,91 @@
+  });
+
+
+  // Same test as propagation-stackdriver.
```
This seems to suggest that we don't really need `generate` to exist as a function on `Propagation` mechanisms. Not really anything to do in this PR, but what do you think?

tcolgate replied:
I couldn't see where that function actually every got called from. I've updated the jaeger version to produce 64 bit spanIds. I think the exporter gets to generate brand new traces/spans maybe? I don't think the propagator is involved at all.

kjin commented on packages/opencensus-propagation-jaeger/tsconfig.json:
```diff
@@ -0,0 +1,18 @@
+    "pretty": true,
+    "module": "commonjs",
+    "target": "es6",
+    "strictNullChecks": false
```
Not sure if this `tsconfig.json` was cloned from somewhere else, but either way, would the code still compile if this option was removed?

tcolgate commented:
@kjin 
- Fixed a couple of errors in README
- Added link to jaeger client docs in the source (there was already one in the readme)
- Fixed things to work with strict null checks
- Updated traceId generation to use a 64 bit spanId

kjin commented on packages/opencensus-propagation-jaeger/src/jaeger-format.ts:
```diff
@@ -0,0 +1,105 @@
+    return {
+      traceId: uuid.v4().split('-').join(''),
+      // tslint:disable-next-line:ban Needed to parse hexadecimal.
+      spanId: parseInt(crypto.randomBytes(8).toString('hex'), 16).toString(),
```
Just a warning here -- we used 6 bytes in other `generate()` functions because not all 8-byte unsigned integers are representable in JS. It probably won't have any end-user visible effect but `crypto.randomBytes(6)` has a better guarantee of uniformity.

tcolgate replied:
Cool, I'll revert to six as it also verifies the padding.

kjin commented on packages/opencensus-propagation-jaeger/test/test-jaeger-format.ts:
```diff
@@ -0,0 +1,90 @@
+
+import {JaegerFormat} from '../src/';
+
+const UBER_TRACE_ID = 'uber-trace-id';
```
Can you change this to `TRACE_ID_HEADER` as well.

kjin commented:
Thanks!
